src/main/java/com/aws/greengrass/deployment/DeploymentQueue.java

```diff
@@ -6,76 +6,162 @@
+    private static final String DISCARDED_DEPLOYMENT_ID_LOG_KEY = "DiscardedDeploymentId";
     private static final Logger logger = LogManager.getLogger(DeploymentQueue.class);
-    private final ConcurrentLinkedQueue<Deployment> deploymentsQueue = new ConcurrentLinkedQueue();
 
     /**
-     * Add a deployment to the queue.
+     * Internal queue of deployment id's. For shadow deployments, the internal queue id is not the same as the
+     * actual deployment id.
+     */
+    private final Queue<String> deploymentIdQueue = new LinkedList<>();
+
+    /**
+     * Map of internal queue id -> deployment instance.
+     */
+    private final Map<String, Deployment> deploymentMap = new HashMap<>();
+
+    /**
+     * <p>If the offered deployment id is unique, then insert the offered deployment at the tail of the queue.</p>
+     *
+     * <p>When an offered deployment has the same deployment id as an already-enqueued element:
+     *     - if the offered deployment meets replacement criteria, then the enqueued element is replaced by the offered
+     *       element, preserving queue order.
+     *     - otherwise, the offered deployment is ignored.</p>
      *
-     * @param deployment deployment
+     * <p>Replacement criteria are as follows:
+     *     - offered.getDeploymentStage().getPriority() > enqueued.getDeploymentStage().getPriority(), OR
+     *     - offered.isCancelled() == true, OR
+     *     - offered.getDeploymentType().equals(SHADOW)</p>
+     *
+     * <p>When an offered deployment is a shadow deployment:
+     *     - if there is already a shadow deployment in the queue, then replace it and preserve queue order.
+     *     - otherwise, enqueue normally.</p>
+     *
+     * @param offeredDeployment the offered deployment instance.
+     * @return true if the queue was modified, otherwise false.
      */
-    public synchronized boolean offer(Deployment deployment) {
-        //For shadow deployment when desired state is reverted, it can result in scheduling a deployment which is
-        // same as
-        if (!DeploymentType.SHADOW.equals(deployment.getDeploymentType()) && deploymentsQueue.contains(deployment)) {
+    public synchronized boolean offer(Deployment offeredDeployment) {
+        if (offeredDeployment == null) {
+            return false;
+        }
+        final String offeredDeploymentInternalId; // the id to employ in the internal queue
+        if (Deployment.DeploymentType.SHADOW.equals(offeredDeployment.getDeploymentType())) {
+            offeredDeploymentInternalId = SHADOW_DEPLOYMENT_QUEUE_ID;
+        } else {
+            offeredDeploymentInternalId = offeredDeployment.getId();
+        }
+
+        // is the internal queue id already in use?
+        final String enqueuedDeploymentId = deploymentIdQueue.stream()
+                .filter(enqueuedId -> enqueuedId.equals(offeredDeploymentInternalId))
+                .findAny()
+                .orElse(null);
+        if (enqueuedDeploymentId != null) {
```

Comment:
this seems to defeat a benefit of using an Optional - avoiding nulls

It's perhaps more readable if you made this:

```
Optional<String> enqueuedDeploymentId = ...
if (enqueudDeploymentId.isPresent()) {

}
```

Reply:
Whoops, completely agree :)